src/compartment.c

```diff
@@ -981,47 +1030,74 @@ extract_sym_offset(struct Compartment *comp, struct LibSymSearchResult res)
 {
-    for (size_t i = 0; i < comp_to_search->libs_count; ++i)
+    struct LibSymSearchResult res = { -1, -1, false };
+    if (in_lib)
+    {
+        res = find_lib_dep_sym_in_lib(
+            to_find, comp_to_search, source_lib_idx, sym_type);
+    }
+    else
     {
-        for (size_t j = 0; j < comp_to_search->libs[i]->lib_syms_count; ++j)
+        for (size_t i = 0; i < comp_to_search->libs_count; ++i)
         {
-            // Ignore non-symbol relocations
-            if (!comp_to_search->libs[i]->lib_syms[j].sym_name)
+            if (i == source_lib_idx)
             {
                 continue;
             }
+            res = find_lib_dep_sym_in_lib(to_find, comp_to_search, i, sym_type);
+            if (res.found)
+            {
+                break;
+            };
+        }
+    }
+    return res;
+}
 
-            // TODO eyeball performance of this approach versus using `&&`
-            // Ignore `LOCAL` bind symbols - they cannot be relocated against
-            bool cond
-                = comp_to_search->libs[i]->lib_syms[j].sym_bind != STB_LOCAL;
+static struct LibSymSearchResult
+find_lib_dep_sym_in_lib(const char *to_find, struct Compartment *comp_to_search,
+    const size_t lib_idx, const unsigned short sym_type)
+{
+    const size_t syms_count = comp_to_search->libs[lib_idx]->lib_syms_count;
+    const struct LibDependencySymbol *syms
+        = comp_to_search->libs[lib_idx]->lib_syms;
+    struct LibSymSearchResult res = { -1, -1, false };
+    for (size_t i = 0; i < syms_count; ++i)
+    {
+        struct LibDependencySymbol curr_sym = syms[i];
 
-            // Check symbol name matches
-            cond = cond
-                && !strcmp(
-                    to_find, comp_to_search->libs[i]->lib_syms[j].sym_name);
+        // Ignore no-named symbols
+        if (!curr_sym.sym_name)
+        {
+            continue;
+        }
 
-            // Check symbol type matches
-            cond = cond
-                && comp_to_search->libs[i]->lib_syms[j].sym_type == sym_type;
+        // Ignore `LOCAL` bind symbols - they cannot be relocated against
+        bool cond = curr_sym.sym_bind != STB_LOCAL;
 
-            // Symbols cannot have 0-offset values
-            if (sym_type != STT_TLS)
-            {
-                cond = cond
-                    && comp_to_search->libs[i]->lib_syms[j].sym_offset != 0;
-            }
+        // Check symbol name matches
+        cond = cond && !strcmp(to_find, curr_sym.sym_name);
 
-            if (cond)
-            {
-                struct LibSymSearchResult res = { i, j };
-                return res;
-            }
+        // Check symbol type matches
+        cond = cond && curr_sym.sym_type == sym_type;
+
+        // Symbols cannot have 0-offset values (except if they are a TLS symbol)
+        if (sym_type != STT_TLS)
+        {
+            cond = cond && curr_sym.sym_offset != 0;
+        }
+
+        if (cond)
```

Comment:
Would it be clearer to say `if (cond && curr_sym.sym_type == sym_type && curr_sym.sym_offset != 0)` ?

Reply:
I prefer it this way, since then I have each condition separate, with a clarifying comment. I would assume the compiler will collapse it down to something better.

Reply:
I found this really hard to read, because it's so unconventional, and it took me a while to convince myself it was intentional. Another way around this is to use a nested `if` with a comment at each level?

Reply:
As per our offline talk, I've added some additional clarifying comments about the use of `cond` here - https://github.com/capablevms/CHERI-ELF-comp/pull/33/commits/faf4545698fcfe6a8769b9c4642d319278696563.